### PR TITLE
[ResourceIsolation] fe deliver WorkGroups to be

### LINKED ISSUE #2848: FE support CRUD of WorkGroup in FE
## Enhancement

WorkGroup is a concept that describe resource isolation in query evaluation, queries are classified into several perdefined workgroups according to its origination information, such as:
1. user, who issue this query?
2. role, which role the user belongs to?
3. query_type, what type is this query?
4. source_ip, from where this query is issued?

The queries' origination information is defined as workgroup classifier.

A User can manipulate a WorkGroup as follows.
1. create a workgroup.
2. drop a workgroup
3. alter a workgroup
4. show all workgroup
5. describe a workgroup

Notice: workgroup is internal term, for users perception, resource_group is used instead.

### LINKED ISSUE #3450: FE delivers WorkGroup to BE
## Feature request

Workgroup concept, to see https://github.com/StarRocks/starrocks/issues/2848
Workgroup stmt syntax parsing, to see https://github.com/StarRocks/starrocks/issues/3440
Workgroup stmt ast analyzing, to see https://github.com/StarRocks/starrocks/issues/3445
Workgroup stmt processing, to see https://github.com/StarRocks/starrocks/issues/3449

WorkGroups are maintained in FE, it should be delivered to BE to limit associative queries resource commitent. There two methods to transfer WorkGroup information from FE to BE.
1. When FE delivers fragment instances to BE, the corresponding WorkGroup is carried in request of RPC exec_plan_fragment.
2. BE report its workgroup information to FE  in every 5 seconds, FE send WorkGroups in the replies.

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

https://github.com/StarRocks/starrocks/issues/3450
https://github.com/StarRocks/starrocks/issues/2848

### CODE REVIEW

**Review by liuyehcf — DISMISSED** [review]

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 1]
```diff
@@ -278,7 +284,9 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
     }
 
     _query_ctx->fragment_mgr()->register_ctx(fragment_instance_id, std::move(fragment_ctx));
-
+    for (auto& driver : _fragment_ctx->drivers()) {
+        driver->set_workgroup(wg);
+    }
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/fragment_executor.cpp:289]: `set_workgroup()` has already been invoked in L281.

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 2]
```diff
@@ -1950,6 +1963,18 @@ public FragmentExecParams(PlanFragment fragment) {
...
+                        if (workgroupId > 0) {
+                            TWorkGroup wg = new TWorkGroup();
+                            wg.setName("");
+                            wg.setId(ConnectContext.get().getSessionVariable().getWorkGroupId());
+                            wg.setVersion(0);
+                            params.setWorkgroup(wg);
+                        } else if (workgroup != null) {
+                            params.setWorkgroup(workgroup.toThrift());
+                        }
+
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1977]: Remove `TODO` in L1982.

**Review by liuyehcf — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

**Review by liuyehcf — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/master/ReportHandler.java [hunk 3]
```diff
@@ -372,6 +387,18 @@ private static void diskReport(long backendId, Map<String, TDisk> backendDisks)
                 backendId, (System.currentTimeMillis() - start));
     }
 
+    private static void workgroupReport(long backendId, List<TWorkGroup> workGroups) {
+        LOG.info("begin to handle workgroup report from backend{}", backendId);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/master/ReportHandler.java:391]:
```suggestion
        LOG.debug("begin to handle workgroup report from backend{}", backendId);
```

File: fe/fe-core/src/main/java/com/starrocks/master/ReportHandler.java [hunk 4]
```diff
@@ -372,6 +387,18 @@ private static void diskReport(long backendId, Map<String, TDisk> backendDisks)
...
+        long start = System.currentTimeMillis();
+        Backend backend = Catalog.getCurrentSystemInfo().getBackend(backendId);
+        if (backend == null) {
+            LOG.warn("backend does't exist. id: " + backendId);
+        }
+        Catalog.getCurrentCatalog().getWorkGroupMgr().saveActiveWorkGroupsForBe(backendId, workGroups);
+        LOG.info("finished to handle workgroup report from backend{}, cost: {} ms, num: {}",
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/master/ReportHandler.java:398]:
```suggestion
        LOG.debug("finished to handle workgroup report from backend{}, cost: {} ms, num: {}",
```

File: fe/fe-core/src/main/java/com/starrocks/master/ReportHandler.java [hunk 5]
```diff
@@ -372,6 +387,18 @@ private static void diskReport(long backendId, Map<String, TDisk> backendDisks)
                 backendId, (System.currentTimeMillis() - start));
     }
 
+    private static void workgroupReport(long backendId, List<TWorkGroup> workGroups) {
+        LOG.info("begin to handle workgroup report from backend{}", backendId);
+        long start = System.currentTimeMillis();
+        Backend backend = Catalog.getCurrentSystemInfo().getBackend(backendId);
+        if (backend == null) {
+            LOG.warn("backend does't exist. id: " + backendId);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/master/ReportHandler.java:395]: The check only print log?

Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/master/ReportHandler.java:395]: According to diskReport method,  it is a defensive programming style.

**Review by Seaven — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

**Review by liuyehcf — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 6]
```diff
@@ -341,6 +343,9 @@
     @VariableMgr.VarAttr(name = PIPELINE_PROFILE_MODE)
     private String pipelineProfileMode = "brief";
 
+    @VariableMgr.VarAttr(name = WORKGROUP_ID)
+    private int workgroupId = 0;
```
[the remark this answers is not in the document]
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:347]: fixed

**Review by stdpain — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by laotan332 — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Pull request merged**